### ISSUE TITLE
Full screen toggle forgets floating windows size and location

### ISSUE THREAD

**Issue by justinesmithies** [qtile/qtile#4058]
### The issue:

Qtile version: 0.22.2.dev80+g9e16cb5b

I have a hook that when mpv opens it floats, resizes and positions the mpv window. 

```
@hook.subscribe.client_managed
def repos(window):
    if window.get_wm_class() and 'mpv' in window.get_wm_class():
        window.floating = True
        window.place(1200, 650, 640, 360, 2, "#ffffff")
```

When i use my key bind to full screen toggle the window it works as expected but when I then toggle it again I'd expect it to return to the exact size and location plus floating. It doesn't it just tiles the mpv window.

```
    # Toggle Fullscreen
    Key([mod], "f",
        lazy.window.toggle_fullscreen(),
        lazy.hide_show_bar(position='all'),
        desc='Toggle fullscreen and the bars'
        ),
```

Is this normal behaviour ? As if I use my key bind to move the mpv window to another workspace or monitor works as expected.

### Required:

- [X] I have searched past issues to see if this bug has already been reported.

**Comment by elParaguayo** [comment]
I think that's normal. I don't think we save position/state when toggling fullscreen.

**Comment by justinesmithies** [comment]
Oh OK in that case could it be added as an enhancement as I for one would find it incredibly useful and I'm sure others would too ?

**Comment by m-col** [comment]
I think this also happens when windows are minimised, or at least they forget whether they were floating. We currently track this float state where each window has exactly one of these flags:
```py
@enum.unique
class FloatStates(enum.Enum):
    NOT_FLOATING = 1
    FLOATING = 2
    MAXIMIZED = 3
    FULLSCREEN = 4
    TOP = 5
    MINIMIZED = 6
```
Perhaps this could be improved by instead tracking floating vs tiled as one flag and then the other states as another.

**Comment by jwijenbergh** [comment]
I agree that we need to do a refactor of how fullscreen windows work. Just getting fullscreen windows to properly restore the layout was a hassle to figure out

**Comment by mitchdost** [comment]
Hi,

think I'm also having a reproducable issue that belongs to the same topic. Until 0.22.1 a virtual box machine opens in fullscreen and non-floating mode.

In version 0.23.0 it opens in floating mode and the upper left corner of the window is positioned in the middle of the screen.
May it be possible, that the actual problem is the floating mode?

I attached two files with output from xprop. One for 0.22.1 and one for 0.23.0

[qtile-0.22.1-xprop-virt-windows.txt](https://github.com/qtile/qtile/files/13348546/qtile-0.22.1-xprop-virt-windows.txt)
[qtile-0.23.0-xprop-virt-windows.txt](https://github.com/qtile/qtile/files/13348549/qtile-0.23.0-xprop-virt-windows.txt)
